lib/rules/no-restricted-properties.js

```diff
@@ -142,40 +142,39 @@ module.exports = {
             }
         }
 
-        /**
-         * Checks property accesses in a destructuring assignment expression, e.g. `var foo; ({foo} = bar);`
-         * @param {ASTNode} node An AssignmentExpression or AssignmentPattern node
-         * @returns {undefined}
-         */
-        function checkDestructuringAssignment(node) {
-            if (node.right.type === "Identifier") {
-                const objectName = node.right.name;
-
-                if (node.left.type === "ObjectPattern") {
-                    node.left.properties.forEach(property => {
-                        checkPropertyAccess(node.left, objectName, astUtils.getStaticPropertyName(property));
-                    });
-                }
-            }
-        }
-
         return {
             MemberExpression(node) {
                 checkPropertyAccess(node, node.object && node.object.name, astUtils.getStaticPropertyName(node));
             },
-            VariableDeclarator(node) {
-                if (node.init && node.init.type === "Identifier") {
-                    const objectName = node.init.name;
-
-                    if (node.id.type === "ObjectPattern") {
-                        node.id.properties.forEach(property => {
-                            checkPropertyAccess(node.id, objectName, astUtils.getStaticPropertyName(property));
-                        });
-                    }
+            ObjectPattern(node) {
+                let objectName;
+
+                if (node.parent.type === "VariableDeclarator") {
+                    objectName = node.parent.init && node.parent.init.type === "Identifier" ? node.parent.init.name : "";
+                } else if ((node.parent.type === "AssignmentExpression" || node.parent.type === "AssignmentPattern") && node.parent.right.type === "Identifier") {
+                    objectName = node.parent.right.name;
+                } else {
+                    objectName = "";
                 }
-            },
-            AssignmentExpression: checkDestructuringAssignment,
-            AssignmentPattern: checkDestructuringAssignment
+
+                if (restrictedProperties.has(objectName)) {
+                    node.properties.forEach(property => {
+                        const propertyName = astUtils.getStaticPropertyName(property);
+
+                        checkPropertyAccess(node, objectName, propertyName);
+                    });
+                } else if (globallyRestrictedObjects.has(objectName)) {
+                    const propertyName = astUtils.getStaticPropertyName(node.properties[0]);
+
+                    checkPropertyAccess(node, objectName, propertyName);
+                } else {
+                    node.properties.forEach(property => {
+                        const propertyName = astUtils.getStaticPropertyName(property);
+
+                        checkPropertyAccess(node, "", propertyName);
+                    });
+                }
```

Comment:
Is it necessary to check restrictions at this point? In the case of a globally restricted object, I think it should still report all properties (if we'd like to report only 1 error for the whole pattern, the message would have to be different).

Could we maybe simplify the code by just always doing this:

```js
node.properties.forEach(property => {
    checkPropertyAccess(node, objectName, astUtils.getStaticPropertyName(property));
});
```

Reply:
No, I removed the restrictions here [`205695e`](https://github.com/eslint/eslint/pull/17818/commits/205695ef69d18e0006b125f9f721685ab3a6cc1a).